### ISSUE TITLE
Wrong labels in single function graphs (1 algo)

### ISSUE THREAD

**Issue by brockho** [numbbo/coco#1494]
When looking at the single function graphs for one algorithm:
[pprldmany_f002.pdf](https://github.com/numbbo/coco/files/1248359/pprldmany_f002.pdf)
the labels are not consistent with our usual notations (2-D: plus, 3-D: triangle pointing down, 5-D: star, 10-D: small circle, 20-D: square, 40: diamond.).

**Comment by brockho** [comment]
The same holds for other graphs of the same type, e.g.:
[pprldmany.pdf](https://github.com/numbbo/coco/files/1248376/pprldmany.pdf)

**Comment by nikohansen** [comment]
FTR, our default symbols have the mnemonics 2: a cross is two lines, 3: a triangle has 3 corners, 5: the star has five tips, 10: is a round number, 20: be creative ;-) ...

**Comment by islandblue** [comment]
Examining the graph, I would say the disorder is due to alphabetic sorting of the dimensions. Something like...

```
HUMAN SORT |  MARKER  | ALPHA SORT
       2-D |   plus   | 10-D
       3-D | triangle | 2-D
       5-D |   star   | 20-D
      10-D |  circle  | 3-D
      20-D |  square  | 40-D
      40-D | diamond  | 5-D
```

So we probably assigned marker thinking it was sorted by dimension like in the first column, where it was actually sorted by label text like in the second column.

**Comment by brockho** [comment]
A reasonable explanation. Just need to find the code now...

**Comment by nikohansen** [comment]
> I would say the disorder is due to alphabetic sorting of the dimensions. 

nice catch!

**Comment by islandblue** [comment]
Observe `i` being assigned here

https://github.com/numbbo/coco/blob/c94b0a0d81d37590cc955f72720fae8b9daddbf4/code-postprocessing/cocopp/compall/pprldmany.py#L713-L715

Observe `i` being used here to select marker

https://github.com/numbbo/coco/blob/c94b0a0d81d37590cc955f72720fae8b9daddbf4/code-postprocessing/cocopp/compall/pprldmany.py#L731

The content of `plotting_style.algorithm_list` in L715 is `['2-D', '3-D', '5-D', '10-D', '20-D', '40-D']`, which, when sorted, would sort alphabetically. So for `i = 0`, we would have `alg = 10-D`. The same `i` goes and selects `'+'` from `genericsettings.dim_related_markers` at L731.

I am pretty convinced there is nothing to be gained by sorting anything in the weird `['10-D', '2-D', '20-D', '3-D', '40-D', '5-D']` order, I would venture to say L715 needs correcting. Without knowledge of original intent in using sorted() there (repeatedly), all I can suggest is to remove sorting altogether. Please advise.

**Comment by brockho** [comment]
The bug is most likely actually my fault, I guess (sorry!), and due to the changes I did to make the postprocessing deterministic when the pseudorandom number generator seed gets fixed. The easiest fix (but probably not the most intuitive to read and understand afterwards) is to change the order of the markers in `genericsettings.dim_related_markers` to the sorting, we observe (`'10-D, '2-D' '20-D', ...`).

But I am open for discussion.

**Comment by islandblue** [comment]
I'm inclined to look for a better solution, because...

- if genericsettings.dim_related_markers is used anywhere else, this may break things elsewhere
- if it's used only here, why not make it a dictionary and absolve ourselves from ordering confusion

**Comment by nikohansen** [comment]
> The easiest fix (but probably not the most intuitive to read and understand afterwards) is to change the order of the markers in genericsettings.dim_related_markers to the sorting

not an option, IMHO ;-)

For the moment I don't understand why the sorting in L715 is needed to make the code deterministic, given that `plotting_style.algorithm_list` is what it promises to be (and what it apparently is): a `list`. The non-determinism problem occurs only with `dict`s. So it should be fine to replace L715 with the simplified version
```python
for i, alg in enumerate(plotting_style.algorithm_list):
```

**Comment by nikohansen** [comment]
FWIW, if the strings obey the `'number-D'` pattern
```python
enumerate(sorted(plotting_style.algorithm_list, key=lambda s: int(s.split('-')[0]))
```
should give the correct sorting irrespectively.

**Comment by islandblue** [comment]
> The original problem should be solved where it actually originates, where a plotting_style.algorithm_list is generated from a dict without taking care of its ordering. This is apparently not the case in this call of the function, but possibly/probably in other calls.

What @nikohansen said is a statement after my own heart. @brockho, can you comment on whether you still think sorting is needed here and maybe hint as to why? Otherwise I'm for targeting where the list is generated and removing all sorting from L715.

**Comment by brockho** [comment]
As I said earlier, I added the sorting to get the postprocessing deterministic if the seed is fixed. But maybe I was not paying attention in this particular case whether this has the desired effect.

I tested now what happens if we take back L715 to `for i, alg in enumerate(plotting_style.algorithm_list):` as suggested by @nikohansen and it works :-) We have the deterministic output over different python versions (tested with Anaconda python 2.7 and 3.4) as well as the right labels.

New version now available in pull request #1524.
